celery/utils/functional.py

```diff
@@ -190,22 +204,40 @@ def __length_hint__(self):
         return self.__it.__length_hint__()
 
     def __iter__(self):
-        return chain(self.__consumed, self.__it)
+        for x in self.__consumed:
+            yield x
+        if not self.__done:
+            for x in self.__it:
+                self.__consumed.append(x)
+                yield x
+            self.__done = True
 
     def __getitem__(self, index):
         if index < 0:
             return self.data[index]
         try:
             return self.__consumed[index]
         except IndexError:
+            it = iter(self)
             try:
                 for _ in range(self.__index, index + 1):
-                    self.__consumed.append(next(self.__it))
+                    next(it)
             except StopIteration:
                 raise IndexError(index)
             else:
                 return self.__consumed[index]
 
+    def __bool__(self):
+        if len(self.__consumed):
+            return True
+
+        try:
+            next(iter(self))
```

Comment:
Isn't this line going to be a problem?
We'll lose an item while checking if it exists.
Shouldn't we also add it to `self. __consumed`?

Reply:
Added a test to make sure we aren't losing any items: https://github.com/celery/celery/pull/6576/commits/f0d4f2e8915dd5b2710c372c5454218ab89e08db.  That next call drops into `self.__iter__` which will add to `self.__consumed`.